File: codes/matlab/008_operator_ternary.m
function stringifyresult = stringify(parameter)
    if isempty(parameter)
        stringifyresult = "null";
        return;
    end
    if isstring(parameter)
        stringifyresult = strcat("""", parameter, """");
        return;
    end
    if ischar(parameter)
        stringifyresult = strcat("""", parameter, """");
        return;
    end
    if isnumeric(parameter)
        stringifyresult = num2str(parameter);
        return;
    end
    if (islogical(parameter) && parameter == true)
        stringifyresult = "true";
        return;
    end
    if (islogical(parameter) && parameter == false)
        stringifyresult = "false";
        return;
    end
    if iscell(parameter)
        if isempty(parameter)
            stringifyresult = "[]";
            return;
        end
        currentresult = "[";
        for arrayitemindex = (1:1:numel(parameter))
            arrayitem = parameter{arrayitemindex};
            currentresult = strcat(currentresult, stringify(arrayitem));
            if (arrayitemindex ~= numel(parameter))
                currentresult = strcat(currentresult, ", ");
            end
        end
        stringifyresult = strcat(currentresult, "]");
        return;
    end
    if isstruct(parameter)
        if (numel(fieldnames(parameter)) == 0)
            stringifyresult = "{}";
            return;
        end
        currentresult = "{";
        objectkeys = fieldnames(parameter);
        for entryindex = (1:1:numel(objectkeys))
            objectkey = objectkeys{entryindex};
            objectvalue = parameter.(objectkey);
            currentresult = strcat(currentresult, """", objectkey, """: ", stringify(objectvalue));
            if (entryindex ~= numel(objectkeys))
                currentresult = strcat(currentresult, ",");
            end
        end
        stringifyresult = strcat(currentresult, "}");
        return;
    end
    stringifyresult = "null";
end

function sprint(varargin)
    currentresult = "";
    for parameterindex = (1:1:numel(varargin))
        parameter = varargin{parameterindex};
        if (iscell(parameter) && (numel(parameter) == 1))
            parameternew = stringify(parameter{1});
            parameternew = strrep(parameternew, """{", "{");
            parameternew = strrep(parameternew, """[", "[");
            parameternew = strrep(parameternew, "}""", "}");
            parameternew = strrep(parameternew, "]""", "]");
            currentresult = strcat(currentresult, parameternew);
            continue;
        end
        currentresult = strcat(currentresult, parameter);
    end
    disp(currentresult);
end

sprint(sprintf("\n"), "% Ternary Operator in Matlab");

% There's no Ternary Operator in Matlab.
% But, we can create our own function to mimic it in Matlab.

function ternaryresult = ternary(truecondition, valueifconditionistrue, valueifconditionisfalse)
    if (islogical(truecondition) && truecondition == true)
        ternaryresult = valueifconditionistrue;
        return;
    end
    ternaryresult = valueifconditionisfalse;
end

CORRECTANSWER = "foo";
sprint("correct answer: ", {CORRECTANSWER});

myanswer = "bar";
sprint("my answer: ", {myanswer});
sprint("is my answer correct: ", {ternary((myanswer == CORRECTANSWER), "correct!", "wrong!")});
% is my answer correct: wrong!

myanswer = "foo";
sprint("my answer: ", {myanswer});
sprint("is my answer correct: ", {ternary((myanswer == CORRECTANSWER), "correct!", "wrong!")});
% is my answer correct: correct!

File: codes/go/008_operator_ternary/008_operator_ternary.go
package main

import (
	"encoding/json"
	"fmt"
)

const EMPTY_STRING = ""
const TAB = "    "

// type any interface{}
type array []any
type object map[string]any

func prettyJsonStringify(anything any) string {
	marshalledJson, err := json.MarshalIndent(anything, EMPTY_STRING, TAB)
	if (err == nil) {
		return string(marshalledJson)
	}
	return "undefined"
}

func ternary(trueCondition bool, valueIfConditionIsTrue any, valueIfConditionIsFalse any) any {
	if (trueCondition == true) {
		return valueIfConditionIsTrue
	}
	return valueIfConditionIsFalse
}

func main() {
	fmt.Println("\n// Ternary Operator in Go")

	const CORRECT_ANSWER = "foo"
	fmt.Println("correct answer:", prettyJsonStringify(CORRECT_ANSWER))

	var myAnswer string

	fmt.Println("// using Ternary Operator function \"ternary\"")

	myAnswer = "bar"
	fmt.Println("my answer: \"" + myAnswer + "\"")
	fmt.Println("is my answer correct:", ternary((myAnswer == CORRECT_ANSWER), "correct!", "wrong!"))
	// is my answer correct: wrong!

	myAnswer = "foo"
	fmt.Println("my answer: \"" + myAnswer + "\"")
	fmt.Println("is my answer correct:", ternary((myAnswer == CORRECT_ANSWER), "correct!", "wrong!"))
	// is my answer correct: correct!
}
